### PR TITLE
Add featurizer

### PR DESCRIPTION
**Pull request by naik-aakash**
Enhancement  > Featurizer for ML #96  
## Added featurizers for ML 

# Todo
- [x]  Add FeaturizeLobsterpy class for lightweight jsons (of manuscript)
- [x] Add FeaturizeCOXX class (fingerprints, w_icohp, moment features)
- [x] Refactor FeaturizeCOXX  class methods
- [x] Add FeaturizeCharges class (Ionicity)
- [x] Add tests for FeaturizeLobsterpy 
- [x] Update test of FeaturizeLobsterpy  to also check exception case
- [x] Add tests for FeaturizeCOXX 
- [x] Add tests for FeaturizeCharges

### CODE REVIEW

**Comment by JaGeo** [comment]
@naik-aakash , I will start here.

File: lobsterpy/featurize/core.py [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+        get_df: returns a pandas dataframe with relevant icohp statistical data as columns from
+        lobsterpy automatic bonding analysis
+
+    """
+
+    def __init__(
+        self,
+        path_to_lobster_calc: str | None = None,
+        path_to_json: str | None = None,
+        bonds: str = "all_bonds",
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:47]: Strings are not consistent with the rest of the code. I think we use "all" and "cation-anion" elsewhere.

File: lobsterpy/featurize/core.py [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+        """
+        if self.path_to_json and not self.path_to_lobster_calc:
+            # read the lightweight lobster json files using read_lobster_lightweight_json method
+            data = self._read_lobster_lightweight_json()
+            if not ids:
+                ids = Path(self.path_to_json).name.split(".")[0]
+
+        elif self.path_to_lobster_calc and not self.path_to_json:
+            # get lobsterpy condensed bonding analysis data using get_lobsterpy_cba_dict method
+            data = self._get_lobsterpy_cba_dict()
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:70]: As far as I had seen it, this part is not yet covered with a test

File: lobsterpy/featurize/core.py [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+                "Please provide either path to lightweight lobster jsons or path to lobster calc"
+            )
+        # define a pandas dataframe
+        df = pd.DataFrame(index=[ids])
+
+        icohp_mean = []
+        icohp_sum = []
+        bond = []
+        antibond = []
+        # extract lobsterpy icohp related data for bond type specified
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:86]: Could you add a comment on what your statistical data means here? You might also want to mention that there will be differences between "all" and "cation-anion" mode, as in "all", the bonds will come up twice.

File: lobsterpy/featurize/core.py [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+        with gzip.open(str(self.path_to_json), "rb") as f:
+            data = json.loads(f.read().decode("utf-8"))
+
+        lobster_data = {}
+        for item in data:
+            lobster_data.update(item)
+
+        return lobster_data
+
+    def _get_lobsterpy_cba_dict(self) -> dict:
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:146]: test needed

File: lobsterpy/featurize/core.py [hunk 5]
```diff
@@ -0,0 +1,829 @@
...
+        lobster lightweight json.gz files
+
+        Returns:
+            Returns a pandas dataframe with lobsterpy icohp statistics
+
+        """
+        if self.path_to_json and not self.path_to_lobster_calc:
+            # read the lightweight lobster json files using read_lobster_lightweight_json method
+            data = self._read_lobster_lightweight_json()
+            if not ids:
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:65]: test needed

File: lobsterpy/featurize/core.py [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+            for k, v in data[self.bonds]["lobsterpy_data"]["sites"].items():
+                if v["bonds"]:
+                    for k1, v1 in v["bonds"].items():
+                        icohp_mean.append(float(v1["ICOHP_mean"]))
+                        icohp_sum.append(float(v1["ICOHP_sum"]))
+                        bond.append(v1["bonding"]["perc"])
+                        antibond.append(v1["antibonding"]["perc"])
+        else:
+            raise Exception(
+                "There exist no data for {} for {} structure. "
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:97]: "There is " ... 
And test needed

File: lobsterpy/featurize/core.py [hunk 7]
```diff
@@ -0,0 +1,829 @@
...
+        cohpcar_path = dir_name / "COHPCAR.lobster.gz"
+        charge_path = dir_name / "CHARGE.lobster.gz"
+        structure_path = dir_name / "POSCAR.gz"
+        icohplist_path = dir_name / "ICOHPLIST.lobster.gz"
+        madelung_energies_path = dir_name / "MadelungEnergies.lobster.gz"
+
+        if self.bonds == "all_bonds":
+            which_bonds = "all"
+        else:
+            which_bonds = "cation-anion"
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:165]: not covered yet

File: lobsterpy/featurize/core.py [hunk 8]
```diff
@@ -0,0 +1,829 @@
...
+
+        if (
+            icohplist_path.exists()
+            and cohpcar_path.exists()
+            and charge_path.exists()
+            and structure_path.exists()
+        ):
+            try:
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:173]: this statement is not yet covered.

File: lobsterpy/featurize/core.py [hunk 9]
```diff
@@ -0,0 +1,829 @@
...
+                analyse = Analysis(
+                    path_to_poscar=str(structure_path),
+                    path_to_icohplist=str(icohplist_path),
+                    path_to_cohpcar=str(cohpcar_path),
+                    path_to_charge=str(charge_path),
+                    summed_spins=False,  # we will always use spin polarization here
+                    cutoff_icohp=0.10,
+                    whichbonds=which_bonds,
+                )
+                data = {}
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:183]:
```suggestion
                data = {self.bonds: {"lobsterpy_data": analyse.condensed_bonding_analysis}}
```

File: lobsterpy/featurize/core.py [hunk 10]
```diff
@@ -0,0 +1,829 @@
...
+                data.update(
+                    {self.bonds: {"lobsterpy_data": analyse.condensed_bonding_analysis}}
+                )
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:186]:
```suggestion
                data = {self.bonds: {"lobsterpy_data": analyse.condensed_bonding_analysis}}
```

File: lobsterpy/featurize/core.py [hunk 11]
```diff
@@ -0,0 +1,829 @@
...
+            except ValueError:
+                data = {}
+                data.update({self.bonds: {"lobsterpy_data": {}}})
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:189]: Not covered by tests

File: lobsterpy/featurize/core.py [hunk 12]
```diff
@@ -0,0 +1,829 @@
...
+            )
+
+            madelung_energies = {
+                "Mulliken": np.nan,
+                "Loewdin": np.nan,
+                "Ewald_splitting": np.nan,
+            }
+            data.update({"madelung_energies": madelung_energies})
+
+        return data
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:219]: Test needed

File: lobsterpy/featurize/core.py [hunk 13]
```diff
@@ -0,0 +1,829 @@
...
+        if max_e is None:
+            max_e = np.max(energies)
+
+        if min_e is None:
+            min_e = np.min(energies)
+
+        if label_list:
+            coxxcar_obj = coxxcar_obj.get_summed_cohp_by_label_list(
+                label_list
+            ).get_cohp()
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:325]: Not covered.

File: lobsterpy/featurize/core.py [hunk 14]
```diff
@@ -0,0 +1,829 @@
...
+                coxx_all = coxxcar_obj[Spin.down]
+            else:
+                raise ValueError(
+                    'LOSBTER calculation is non-spin polarized. Please switch spin_type to "up"'
+                )
+        elif spin_type == "summed":
+            if Spin.down in coxxcar_obj:
+                coxx_all = coxxcar_obj[Spin.up] + coxxcar_obj[Spin.down]
+            else:
+                coxx_all = coxxcar_obj[Spin.up]
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:342]: tests

File: lobsterpy/featurize/core.py [hunk 15]
```diff
@@ -0,0 +1,829 @@
...
+
+            df.at[ids, "COXX_FP"] = fp
+
+            return df
+
+        except KeyError:
+            raise ValueError(
+                "Please recheck fp_type requested argument.Possible options are bonding/antibonding/overall"
+            )
+
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:427]: Not yet covered by tests

File: lobsterpy/featurize/core.py [hunk 16]
```diff
@@ -0,0 +1,829 @@
...
+            bonding_indices = coxx_bf >= 0
+            antibonding_indices = coxx_bf <= 0
+            bnd = abs(trapezoid(coxx_bf[bonding_indices], en_bf[bonding_indices]))
+            antibnd = abs(
+                trapezoid(coxx_bf[antibonding_indices], en_bf[antibonding_indices])
+            )
+            per_bnd = (bnd / (bnd + antibnd)) * 100
+            per_antibnd = (antibnd / (bnd + antibnd)) * 100
+
+        return per_bnd, per_antibnd, w_icoxx, ein
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:498]: tests

File: lobsterpy/featurize/core.py [hunk 17]
```diff
@@ -0,0 +1,829 @@
...
+    ) -> Tuple[float, float, float, float]:
+        if label_list:
+            coxxcar = self.completecoxx.get_summed_cohp_by_label_list(
+                label_list
+            ).get_cohp()
+            try:
+                coxx_all = coxxcar[Spin.up] + coxxcar[Spin.down]
+            except KeyError:
+                coxx_all = coxxcar[Spin.up]
+            energies = self.completecoxx.energies - self.completecoxx.efermi
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:511]: test

File: lobsterpy/featurize/core.py [hunk 18]
```diff
@@ -0,0 +1,829 @@
...
+            coxx_dict["overall"] = coxx_all
+        else:
+            coxx_dict["antibonding"] = np.array(
+                [scohp if scohp <= 0 else 0 for scohp in coxx_all]
+            )
+            coxx_dict["bonding"] = np.array(
+                [scohp if scohp >= 0 else 0 for scohp in coxx_all]
+            )
+            coxx_dict["overall"] = coxx_all
+
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:537]: test

File: lobsterpy/featurize/core.py [hunk 19]
```diff
@@ -0,0 +1,829 @@
...
+        for i, j in enumerate(charges):
+            if (
+                j > 0
+                and not structure.species[i].is_transition_metal
+                and (
+                    not structure.species[i].is_actinoid
+                    and not structure.species[i].is_lanthanoid
+                )
+            ):
+                valence_elec = element(structure.species[i].value)
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:762]: is not tested yet

File: lobsterpy/featurize/core.py [hunk 20]
```diff
@@ -0,0 +1,829 @@
...
+                val = j / (valence_elec.nvalence() - 0)
+                ch_veff.append(val)
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:764]: This is not yet covered. You can also define a method/function for this.

File: lobsterpy/featurize/core.py [hunk 21]
```diff
@@ -0,0 +1,829 @@
...
+            elif (
+                j > 0
+                and not structure.species[i].is_transition_metal
+                and (
+                    not structure.species[i].is_actinoid
+                    and not structure.species[i].is_lanthanoid
+                )
+            ):
+                valence_elec = element(structure.species[i].value)
+                val = j / (valence_elec.nvalence() - 8)
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:775]: Very similar code

Comment (JaGeo) [reply lobsterpy/featurize/core.py:775]: Not yet tested

File: lobsterpy/featurize/core.py [hunk 22]
```diff
@@ -0,0 +1,829 @@
...
+    def get_df(self, ids: str | None = None) -> pd.DataFrame:
+        """
+        This function returns a pandas dataframe with computed ionicity as column
+
+        Returns:
+            Returns a pandas dataframe with ionicity
+
+        """
+        if ids:
+            df = pd.DataFrame(index=[ids])
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:819]: no covered

**Comment by JaGeo** [comment]
@naik-aakash I went to one of the files. It only has 65% coverage (core.py). The other one has only 56%. I would suggest that you increase test coverage first. The code will otherwise easily get bugs in the future. Please reuse test data as good as possible and refactor to only test functions instead of huge code blocks. The latter should help with reducing requirements for new test data.

**Comment by naik-aakash** [comment]
> @naik-aakash I went to one of the files. It only has 65% coverage (core.py). The other one has only 56%. I would suggest that you increase test coverage first. The code will otherwise easily get bugs in the future. Please reuse test data as good as possible and refactor to only test functions instead of huge code blocks. The latter should help with reducing requirements for new test data.



> @naik-aakash I went to one of the files. It only has 65% coverage (core.py). The other one has only 56%. I would suggest that you increase test coverage first. The code will otherwise easily get bugs in the future. Please reuse test data as good as possible and refactor to only test functions instead of huge code blocks. The latter should help with reducing requirements for new test data.



> @naik-aakash I went to one of the files. It only has 65% coverage (core.py). The other one has only 56%. I would suggest that you increase test coverage first. The code will otherwise easily get bugs in the future. Please reuse test data as good as possible and refactor to only test functions instead of huge code blocks. The latter should help with reducing requirements for new test data.



> @naik-aakash I went to one of the files. It only has 65% coverage (core.py). The other one has only 56%. I would suggest that you increase test coverage first. The code will otherwise easily get bugs in the future. Please reuse test data as good as possible and refactor to only test functions instead of huge code blocks. The latter should help with reducing requirements for new test data.

Hi @JaGeo , this problem I also encountered. I should have updated you before. My bad I forgot. This is not the actual coverage and arises coz of use of multiprocessing module. I have fixed it in the workflow of this branch the real coverage is 87% and 91% respectively.

**Comment by naik-aakash** [comment]
```
coverage run -m pytest
coverage combine
coverage report
```
Need to use this 3 commands successively to get correct value of the coverage (Low coverage we are are coz of multiprocessing module)

**Comment by JaGeo** [comment]
I will check again. Thanks.

File: lobsterpy/featurize/core.py [hunk 23]
```diff
@@ -0,0 +1,829 @@
...
+            n: The order for the moment
+            cohp: COXX list
+            energies: energies
+            center: Take moments with respect to the COXX center
+
+        Returns:
+            COXX nth moment in eV
+        """
+        if e_range:
+            coxx = coxx[(energies >= self.e_range[0]) & (energies <= self.e_range[-1])]
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:630]: It probably works but I would still make sure this comparison is okay.

File: lobsterpy/featurize/core.py [hunk 24]
```diff
@@ -0,0 +1,829 @@
...
+        and moment features (center, width, skewness and kurtosis) of COXX in selected energy range
+
+        Returns:
+            Returns a pandas dataframe with cohp/cobi/coop related features as per input file
+
+        """
+        if ids:
+            df = pd.DataFrame(index=[ids])
+        else:
+            ids = os.path.basename(os.path.dirname(self.path_to_coxxcar))
```
Comment (JaGeo) [inline lobsterpy/featurize/core.py:661]: In many parts of the code, we now use "Path". Could you update this?

**Comment by JaGeo** [comment]
Could Ou also add documentation of the new installation process?

File: lobsterpy/featurize/core.py (hunk 24 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:661]: Yes

File: lobsterpy/featurize/core.py [hunk 25]
```diff
@@ -0,0 +1,829 @@
...
+
+        lobster_data = {}
+        for item in data:
+            lobster_data.update(item)
+
+        return lobster_data
+
+    def _get_lobsterpy_cba_dict(self) -> dict:
+        """
+        This function uses lobsterpy.cohp.analyze.Analysis class to generate a python dictionary object
```
[the remark this answers is not in the document]
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:148]: Thanks! Changed it now to a static method.

File: lobsterpy/featurize/core.py (hunk 20 above)
Comment (naik-aakash) [reply lobsterpy/featurize/core.py:764]: I realized, some redundant if-else checks were there, I simplified it now. Hope this should be fine. Am not able to figure out what you meant by having a function/method as  it is already part of method in the code.

**Comment by naik-aakash** [comment]
Hi @JaGeo , I think now the issues with numerical stability should be fixed and also more tests have been added to improve the coverage. If any more changes necessary would be happy to address it 😃

**Comment by naik-aakash** [comment]
Also, I will try to add orbital-wise COXX features (center, width, skew, kurtosis and fingerprints) as well . But will be added in another PR once this is merged.

File: pylintrc [hunk 26]
```diff
@@ -78,67 +72,6 @@ disable=print-statement,
         useless-suppression,
         deprecated-pragma,
         use-symbolic-message-instead,
-        apply-builtin,
```
Comment (JaGeo) [inline pylintrc:81]: @naik-aakash any reason why we have to remove all of them or is this an issue with not having the latest main branch merged?

Comment (naik-aakash) [reply pylintrc:81]: Ah I removed them as they seem deprecated and made the lint workflow print a whole lot of warnings,

**Review by JaGeo — APPROVED** [review]
I looked throught the main implementations. For the featurization of the COHPs and further features, we might want to run some more tests later on.

**Pull request merged**